### PR TITLE
Pass content and css as raw data

### PR DESCRIPTION
**Pull request by jsnanigans**
## Proposed changes

The Idea is that I can pass HTML or whatever content directly instead of just the file path.

## Types of changes

Updated functions that gets CSS content and HTML selectors.
Created extractRawSelector

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

Im not sure if the syntax is acceptable, if you pass a array of strings in `content` and `css` it will behave like before, but if its a object instead of a array, for `content` you pass `raw` with the eg. HTML and `extension` to make the extractors work. For `css` you just pass a object with the css in `raw`

like this:
```javascript
const purgecss = new Purgecss({
  content: [{
    raw: '<span class="stay"></span>',
    extension: 'html'
  }],
  css: [{
    raw: '.stay {display: inline}; .remove {display: block}'
    }]
})
```

### CODE REVIEW

**Comment by Ffloriel** [comment]
Thanks for the PR. This is an interesting feature.
I was thinking of a way to use purgecss to remove unused css with individual vue component. Being able to select the raw content + the extensions could help with this task.

What use cases are you thinking about for this?

File: __tests__/purgecssDefault.test.js [hunk 1]
```diff
@@ -323,4 +323,21 @@ describe('purge methods with files and default extractor', () => {
...
+describe('purge methods with raw content and default extractor', () => {
+    it('remove .remove - content passed', () => {
+        const purgecss = new Purgecss({
+            content: [{
+                raw: '<span class="stay"></span>',
+                extension: 'html'
+            }],
+            css: [{
+                raw: '.stay {display: inline}; .remove {display: block}'
+            }]
```
Comment (Ffloriel) [inline __tests__/purgecssDefault.test.js]: give a valid css for the test, with `;` after `inline`

File: __tests__/purgecssDefault.test.js [hunk 2]
```diff
@@ -323,4 +323,21 @@ describe('purge methods with files and default extractor', () => {
...
+        })
+        const result = purgecss.purge()[0].css
+        expect(result.includes('remove')).toBe(false)
+    })
```
Comment (Ffloriel) [inline __tests__/purgecssDefault.test.js]: add assertion to test that used css are present, like `.stay`

File: src/index.js [hunk 3]
```diff
@@ -82,17 +82,53 @@ class Purgecss {
...
-            files.push({
+        const sources = this.getCssContents(this.options.css, new Set([...cssFileClasses, ...cssRawClasses]))
+
+        return sources
+    }
+
+    getCssContents(cssOptions: Array<any>, cssClasses: Set<string>) {
+        const sources = []
+        cssOptions.forEach(option => {
+            let file = null
```
Comment (Ffloriel) [inline src/index.js]: use `for...of` instead of `forEach`, it stays consistent with `extractFileSelector`

File: src/index.js [hunk 4]
```diff
@@ -82,17 +82,53 @@ class Purgecss {
...
+
+    /**
+     * Extract the selectors present in the passed string using a purgecss extractor
+     * @param {array} content Array of content
+     * @param {array} extractors Array of extractors
+     */
+    extractRawSelector(content: Array<any>, extractors?: Array<ExtractorsObj>): Set<string> {
+        let selectors = new Set()
+
+        content.forEach(option => {
```
Comment (Ffloriel) [inline src/index.js]: use `for...of` instead of `forEach`

File: src/index.js [hunk 5]
```diff
@@ -82,17 +82,53 @@ class Purgecss {
...
+            const content = option.raw
+            const extractor = this.getFileExtractor(`.${option.extension}`, extractors)
+            selectors = new Set([...selectors, ...this.extractSelectors(content, extractor)])
```
Comment (Ffloriel) [inline src/index.js]: use destructuring to get option properties

**Comment by jsnanigans** [comment]
Thanks for the feedback, the for..of is new to me :)

I am using this in a custom plugin for webpack, the part of it that uses purgecss just gets the chunks and assets and purges the css, then replaces the asset with the purged css, this all happens before any files are written.

**Comment by jsnanigans** [comment]
I just noticed that if you accept this merge you could remove `options.stdin` since it does almost the same thing, but with this PR it would be possible to mix files and standard input for content and css

**Comment by GiraffeKey** [comment]
Please add an example of this to the usage section, I just wrote all this:
```js
const temp = path.resolve(__dirname, "../temp")
fs.mkdirSync(temp)
const fd = fs.openSync(path.join(temp, "temp.css"), "w")
fs.writeSync(fd, css, 0, "utf8")

const purgecss = new PurgeCSS()

css = (await purgecss.purge({
	content: config.purge,
	css: [path.join(temp, "temp.css")],
}))[0].css

fs.rmdirSync(temp, { recursive: true })
```
Only to find this PR immediately after :P
Would be nice if the information were more accessible

**Pull request merged**